metric_learn/nca.py

```diff
@@ -27,33 +58,57 @@ def fit(self, X, y):
     X: data matrix, (n x d)
     y: scalar labels, (n)
     """
+    if self.learning_rate != 'deprecated':
+      warnings.warn('"learning_rate" parameter is not used.'
+                    ' It has been deprecated in version 0.4 and will be'
+                    'removed in 0.5', DeprecationWarning)
+
     X, labels = check_X_y(X, y)
     n, d = X.shape
     num_dims = self.num_dims
     if num_dims is None:
         num_dims = d
+
     # Initialize A to a scaling matrix
     A = np.zeros((num_dims, d))
     np.fill_diagonal(A, 1./(np.maximum(X.max(axis=0)-X.min(axis=0), EPS)))
 
     # Run NCA
-    dX = X[:,None] - X[None]  # shape (n, n, d)
-    tmp = np.einsum('...i,...j->...ij', dX, dX)  # shape (n, n, d, d)
-    masks = labels[:,None] == labels[None]
-    for it in xrange(self.max_iter):
-      for i, label in enumerate(labels):
-        mask = masks[i]
-        Ax = A.dot(X.T).T  # shape (n, num_dims)
-
-        softmax = np.exp(-((Ax[i] - Ax)**2).sum(axis=1))  # shape (n)
-        softmax[i] = 0
-        softmax /= softmax.sum()
-
-        t = softmax[:, None, None] * tmp[i]  # shape (n, d, d)
-        d = softmax[mask].sum() * t.sum(axis=0) - t[mask].sum(axis=0)
-        A += self.learning_rate * A.dot(d)
+    mask = labels[:, np.newaxis] == labels[np.newaxis, :]
+    optimizer_params = {'method': 'L-BFGS-B',
+                        'fun': self._loss_grad_lbfgs,
+                        'args': (X, mask, -1.0),
+                        'jac': True,
+                        'x0': A.ravel(),
+                        'options': dict(maxiter=self.max_iter),
+                        'tol': self.tol
+                        }
+
+    # Call the optimizer
+    opt_result = minimize(**optimizer_params)
 
     self.X_ = X
-    self.A_ = A
-    self.n_iter_ = it
+    self.A_ = opt_result.x.reshape(-1, X.shape[1])
+    self.n_iter_ = opt_result.nit
     return self
+
+  @staticmethod
+  def _loss_grad_lbfgs(A, X, mask, sign=1.0):
+    A = A.reshape(-1, X.shape[1])
+    X_embedded = np.dot(X, A.T)  # (n_samples, num_dims)
+    # Compute softmax distances
+    p_ij = pairwise_distances(X_embedded, squared=True)
+    np.fill_diagonal(p_ij, np.inf)
+    p_ij = np.exp(-p_ij - logsumexp(-p_ij, axis=1)[:, np.newaxis])
+    # (n_samples, n_samples)
+
+    # Compute loss
+    masked_p_ij = p_ij * mask
+    p = np.sum(masked_p_ij, axis=1, keepdims=True)  # (n_samples, 1)
+    loss = np.sum(p)
```

Comment:
Stylistically, I prefer the method call `p.sum()` over the numpy function. But it's not a big deal if you like the other way better.

Reply:
It's true that ``p.sum`` is more concise, I guess I like it more too
I'll make the change